upstairs/src/lib.rs

```diff
@@ -1781,6 +1781,65 @@ impl Buffer {
             .map(|v| *v != 0)
             .zip(self.data.chunks(self.block_size))
     }
+
+    /// Splits the buffer into two at the given index.
+    ///
+    /// Afterwards `self` contains elements `[0, at)`, and the returned `Buffer`
+    /// contains elements `[at, capacity)`.
+    ///
+    /// This is an `O(1)` operation that just increases the reference count and
+    /// sets a few indices.
+    ///
+    /// # Panics
+    /// The `index` must be an even multiple of block size
+    pub fn split_off(&mut self, index: usize) -> Self {
+        assert_eq!(index % self.block_size, 0);
+        let data = self.data.split_off(index);
+        let owned = self.owned.split_off(index / self.block_size);
+        Self {
+            block_size: self.block_size,
+            data,
+            owned,
```

Comment:
I believe using `owned` like this is okay, but we should probably update the comment in `upstairs/src/lib.rs`
for `pub struct Buffer`, as it describes ownership in terms of a block being written to.
As I read that comment though, I wonder if this buffer splitting works with the scrubber correctly?

If the scrubber is set to 131072, then that should fall below the level of breaking up IOs.

Reply:
Fixed the comment in a2e5c68eeeb3a8199a852ca40aba1e10f956a57b (that should have been changed in #1200, oops)

Can you say more about interactions with the scrubber?  I'm not as familiar with how it works.

Reply:
I'm not sure myself if this buffer layer does interact, but I'd like to call @jmpesp to the witness
stand and have his thoughts here.  The [scrubber](https://github.com/oxidecomputer/crucible/blob/6f7884ae4dcf21ee7a92664b83d9b4ae89130223/upstairs/src/volume.rs#L301) just reads from the ROP and writes
to the region, so on the surface I don't see the buffer's owner being used, but I suspect there
might be something under the covers, or in other backends that make use of it?

Reply:
As far as I can tell, I haven't changed any behavior of `owned`.  It sounds like you think something has changed there?

Reply:
Your understanding is correct – we're splitting the buffer into two smaller buffers, but `data` and `owned` remain consistent w.r.t. each other.

Reply:
> The scrubber just reads from the ROP and writes to the region, so on the surface I don't see the buffer's owner being used, but I suspect there might be something under the covers, or in other backends that make use of it?

No, you're right - the scrubber does a `read` (which returns a `Buffer`), then calls `write_unwritten` and passes it `buffer.into_bytes()`, tossing away ownership information. But as I understand it this PR doesn't change the Buffer that would be returned, it simply does the splitting and joining under the covers. The caller of `read` won't notice a difference. If it did, that'd be a bug :)